renku/core/incubation/database.py

```diff
@@ -0,0 +1,671 @@
+        }
+
+    def __setstate__(self, data):
+        self._name = data.pop("name")
+        self._value_type = get_class(data.pop("value_type"))
+        self._key_type = get_class(data.pop("key_type"))
+        self._attribute = data.pop("attribute")
+        self._entries = data.pop("entries")
+
+    @property
+    def name(self) -> str:
+        """Return Index's name."""
+        return self._name
+
+    @property
+    def value_type(self) -> type:
+        """Return Index's value_type."""
+        return self._value_type
+
+    def get(self, key, default=None):
+        """Return an entry based on its key."""
+        return self._entries.get(key, default)
+
+    def values(self):
+        """Return an iterator of values."""
+        return self._entries.values()
+
+    def items(self):
+        """Return an iterator of keys and values."""
+        return self._entries.items()
+
+    def add(self, object: Persistent, *, key_object=None):
+        """Update index with object.
+
+        If ``key_object`` is not None then it is used to generate index keys, otherwise, ``object`` is used as the key.
+        """
+        assert isinstance(object, self._value_type), f"Cannot add objects of type '{type(object)}'"
+
+        key = self._get_key(object=object, key_object=key_object)
+        self._entries[key] = object
+
+    def remove(self, object: Persistent, *, key_object=None):
+        """Delete an object if it is indexed."""
+        assert isinstance(object, self._value_type), f"Invalid object type '{type(object)}'"
+
+        key = self._get_key(object=object, key_object=key_object)
+        self._entries.pop(key, None)
+
+    def _get_key(self, object: Persistent, key_object):
+        if key_object is None:
+            key_object = object
+
+        if self._key_type is not None:
+            assert isinstance(key_object, self._key_type), f"Invalid key type: {type(key_object)} != {self._key_type}"
+
+        return get_attribute(key_object, self._attribute)
+
+
+class IndexList(Index):
+    """Database index pointing to multiple values."""
+
+    def add(self, object: Persistent, *, key_object=None):
+        """Update index with object.
+
+        If ``key_object`` is not None then it is used to generate index keys, otherwise, ``object`` is used as the key.
+        """
+        assert isinstance(object, self._value_type), f"Cannot add objects of type '{type(object)}'"
+
+        key = self._get_key(object=object, key_object=key_object)
+        values = self._entries.get(key)
+        if values:
+            assert isinstance(values, list), f"Value in IndexList is not a list: {values}"
+            for value in values:
+                if value._p_oid == object._p_oid:  # Object is already indexed
+                    return
+            values.append(object)
+            self._entries._p_changed = True
+            # TODO: Do we need self._p_changed = True
+        else:
+            self._entries[key] = [object]
```

Comment:
If we used `PersistentList` (https://persistent.readthedocs.io/en/latest/api/collections.html) we wouldn't have to set _p_changed on append manually, I think.